src/main/scala/chisel3/util/Decoupled.scala

```diff
@@ -215,28 +220,35 @@ class Queue[T <: Data](val gen: T,
     }
   }
 
-  val io = IO(new QueueIO(genType, entries))
-
+  val io = IO(new QueueIO(genType, entries, hasFlush))
   val ram = if (useSyncReadMem) SyncReadMem(entries, genType, SyncReadMem.WriteFirst) else Mem(entries, genType)
   val enq_ptr = Counter(entries)
   val deq_ptr = Counter(entries)
   val maybe_full = RegInit(false.B)
-
   val ptr_match = enq_ptr.value === deq_ptr.value
   val empty = ptr_match && !maybe_full
   val full = ptr_match && maybe_full
   val do_enq = WireDefault(io.enq.fire())
   val do_deq = WireDefault(io.deq.fire())
-
-  when (do_enq) {
-    ram(enq_ptr.value) := io.enq.bits
-    enq_ptr.inc()
-  }
-  when (do_deq) {
-    deq_ptr.inc()
-  }
-  when (do_enq =/= do_deq) {
-    maybe_full := do_enq
+  val flush = io.flush.getOrElse(false.B) 
+
+  // when flush is on, empty the queue
+  // flush is the highest priority operation
+  when(flush) {
+    enq_ptr.reset()
+    deq_ptr.reset()
+    maybe_full := false.B
+  } .otherwise {
+    when (do_enq) {
+      ram(enq_ptr.value) := io.enq.bits
+      enq_ptr.inc()
+    }
+    when (do_deq) {
+      deq_ptr.inc()
+    }
+    when (do_enq =/= do_deq) {
+      maybe_full := do_enq
+    }
   }
```

Comment:
We should get rid of the arc from `flush` to `ram` because we don't care about the contents of `ram` following a flush.  This is especially desirable since this is a high fan-out path.  To do this, avoid the `otherwise` and rewrite as:

      when (do_enq) { 
        ram(enq_ptr.value) := io.enq.bits
        enq_ptr.inc()
      }
      when (do_deq) { 
        deq_ptr.inc()
      }
      when (do_enq =/= do_deq) { 
        maybe_full := do_enq
      }
      when (flush) {
        enq_ptr.reset()
        deq_ptr.reset()
        maybe_full := false.B
      }

Reply:
Okay I implemented it like you suggested! Thanks!